File: frontend/herdr-web/src/components/Assistant/AssistantSession.ts
import { ApiError } from "../../api/client";
import { isTerminalRunStatus, type AgentRun } from "../../api/agentRuns";
import type { AssistantContext, AssistantRequest, AssistantTransport } from "../../api/assistant";

export interface AssistantState {
  draft: string;
  turns: AgentRun[];
  context: AssistantContext;
  pending: AssistantRequest | null;
  busy: boolean;
  ready: boolean;
  error: string | null;
}

/** Feature-independent session. View unmount never stops accepted server work. */
export class AssistantSession {
  private listeners = new Set<() => void>();
  private stopped = false;
  private preparing = false;
  private observing = false;
  state: AssistantState;
  constructor(readonly key: string, readonly paneId: string, readonly rootPath: string | undefined,
              context: AssistantContext, private transport: AssistantTransport) {
    this.state = { draft: "", turns: [], context, pending: null, busy: false, ready: false, error: null };
    try {
      const stored = sessionStorage.getItem(key);
      if (stored) this.state = { ...this.state, ...JSON.parse(stored), busy: false, ready: false, error: null };
    } catch { /* A restricted browser still supports an in-memory conversation. */ }
  }
  subscribe = (listener: () => void) => { this.listeners.add(listener); return () => { this.listeners.delete(listener); }; };
  snapshot = () => this.state;
  private set(patch: Partial<AssistantState>) {
    this.state = { ...this.state, ...patch };
    try { sessionStorage.setItem(this.key, JSON.stringify(this.state)); } catch { /* In-memory fallback. */ }
    this.listeners.forEach((listener) => listener());
  }
  setDraft = (draft: string) => this.set({ draft });
  get latest() { return this.state.turns[this.state.turns.length - 1]; }
  async prepare() {
    if (this.state.ready || this.preparing) return;
    this.preparing = true;
    try {
      const caps = await this.transport.capabilities();
      if (!caps.profiles.includes("contextual-question-v1")) throw new Error("Update this machine's companion to use contextual questions.");
      this.set({ ready: true });
      if (this.latest && !isTerminalRunStatus(this.latest.status)) void this.observe(this.latest.id);
    } catch (error) { this.set({ error: message(error) }); }
    finally { this.preparing = false; }
  }
  async submit() {
    if (!this.state.ready || this.state.busy || this.state.pending || !this.state.draft.trim() ||
        (this.latest && (!isTerminalRunStatus(this.latest.status) || this.latest.status === "promoted"))) return;
    this.set({ pending: { profile: "contextual-question-v1", mode: "ask", prompt: this.state.draft,
      paneId: this.paneId, scope: { expectedRootPath: this.rootPath }, context: this.state.context,
      clientRequestId: crypto.randomUUID(), continueFromRunId: this.latest?.id } });
    await this.reconcile();
  }
  async reconcile() {
    if (this.state.busy || !this.state.pending) return;
    const request = this.state.pending;
    this.stopped = false;
    this.set({ busy: true, error: null });
    try {
      const run = await this.transport.start(request);
      this.update(run);
      this.set({ pending: null, draft: this.state.draft === request.prompt ? "" : this.state.draft });
      if (this.stopped) this.update(await this.transport.stop(run.id));
      await this.observe(run.id);
    } catch (error) {
      if (error instanceof ApiError && [400, 403, 404, 409, 413, 422].includes(error.status)) this.set({ pending: null, error: message(error) });
      else this.set({ error: message(error) + " Reconcile submission before starting another question." });
    }
    finally { this.set({ busy: false }); }
  }
  private update(run: AgentRun) {
    const exists = this.state.turns.some((turn) => turn.id === run.id);
    this.set({ turns: exists ? this.state.turns.map((turn) => turn.id === run.id ? run : turn) : [...this.state.turns, run] });
  }
  async observe(id = this.latest?.id) {
    if (!id || this.observing) return;
    this.observing = true;
    this.set({ busy: true });
    let failures = 0;
    try {
      for (;;) {
        try {
          const run = await this.transport.fetch(id);
          this.update(run);
          this.set({ error: null });
          failures = 0;
          if (isTerminalRunStatus(run.status)) return;
        } catch (error) {
          this.set({ error: "Reconnecting: " + message(error) });
          if (++failures >= 10) return;
        }
        await new Promise((resolve) => setTimeout(resolve, failures ? 3000 : 900));
      }
    } finally { this.observing = false; this.set({ busy: false }); }
  }
  async stop() {
    this.stopped = true;
    if (this.state.pending || !this.latest) return;
    try { this.update(await this.transport.stop(this.latest.id)); }
    catch (error) { this.set({ error: message(error) }); }
  }
  async promote() {
    if (this.state.busy || this.latest?.status !== "completed") return;
    this.set({ busy: true, error: null });
    try { this.update(await this.transport.promote(this.latest.id)); }
    catch (error) { this.set({ error: message(error) }); }
    finally { this.set({ busy: false }); }
  }
  newQuestion() {
    if (this.state.busy || this.state.pending || (this.latest && !isTerminalRunStatus(this.latest.status))) return;
    this.set({ turns: [], draft: "", error: null });
  }
  addContext(item: AssistantContext["items"][number]) {
    if (this.state.busy || this.state.pending) return;
    this.set({ context: { ...this.state.context, snapshotId: crypto.randomUUID(), capturedAt: new Date().toISOString(),
      items: [...this.state.context.items.filter((old) => old.id !== item.id), item] } });
  }
  removeContext(id: string) {
    if (this.state.busy || this.state.pending) return;
    this.set({ context: { ...this.state.context, snapshotId: crypto.randomUUID(), capturedAt: new Date().toISOString(),
      items: this.state.context.items.filter((item) => item.id !== id) } });
  }
}
function message(error: unknown) { return error instanceof Error ? error.message : "Question unavailable"; }

File: herdr-harness-mac/herdr-harness-mac/Assistant/AssistantSession.swift
import Foundation
import Observation

@MainActor @Observable
final class AssistantSession {
    let title: String
    let machineID: String
    let paneID: String?
    let rootPath: String?
    var context: AssistantContext
    var currentContext: AssistantContext
    var draft = ""
    var selectedModel = ""
    private(set) var turns: [HeadlessAgentRun] = []
    private(set) var models: [PiAvailableModel] = []
    private(set) var isRunning = false
    private(set) var isReady = false
    private(set) var error: String?
    private(set) var pending: AssistantRequest?
    private(set) var isPromoting = false
    @ObservationIgnored private let transport: AssistantTransport
    @ObservationIgnored private let persistence: AssistantPersistence
    @ObservationIgnored private var operation: Task<Void, Never>?
    @ObservationIgnored private var stopRequested = false
    @ObservationIgnored private var isPreparing = false

    init(title: String, machineID: String, paneID: String?, rootPath: String?, context: AssistantContext,
         transport: AssistantTransport, persistence: AssistantPersistence) {
        self.title = title
        self.machineID = machineID
        self.paneID = paneID
        self.rootPath = rootPath
        self.context = context
        self.currentContext = context
        self.transport = transport
        self.persistence = persistence
    }

    var latest: HeadlessAgentRun? { turns.last }
    var canSend: Bool { isReady && !isRunning && !isPromoting && pending == nil && latest?.status != .promoted && latest?.status.isTerminal != false }

    func prepare() async {
        guard !isReady, !isPreparing else { return }
        isPreparing = true
        defer { isPreparing = false }
        if let saved = await persistence.load() {
            context = saved.context
            draft = saved.draft
            turns = saved.turns
            pending = saved.pending
            selectedModel = saved.selectedModel
        }
        do {
            let capabilities = try await transport.capabilities()
            guard capabilities.profiles.contains("contextual-question-v1") else {
                error = "Update this machine's companion server to use contextual questions."
                return
            }
            isReady = true
            if let catalog = try? await transport.models() { models = catalog.models }
            if let latest, !latest.status.isTerminal {
                isRunning = true
                operation = Task { await observe(latest.id) }
            }
        } catch { self.error = "Contextual questions are unavailable: \(error.localizedDescription)" }
    }

    func submit() {
        guard canSend, !draft.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else { return }
        pending = AssistantRequest(prompt: draft, paneId: paneID, scope: .init(expectedRootPath: rootPath),
                                   context: context, continueFromRunId: latest?.id,
                                   model: selectedModel.isEmpty ? nil : selectedModel)
        retrySubmission()
    }

    func retrySubmission() {
        guard let request = pending, !isRunning else { return }
        isRunning = true
        stopRequested = false
        error = nil
        operation = Task {
            guard await save() else { isRunning = false; return }
            do {
                let run = try await transport.start(request)
                update(run)
                pending = nil
                if draft == request.prompt { draft = "" }
                _ = await save()
                if stopRequested { try await cancelAccepted(run.id) }
                await observe(run.id)
            } catch {
                if case let APIError.server(status, _) = error, [400, 403, 404, 409, 413, 422].contains(status) {
                    pending = nil
                    self.error = error.localizedDescription
                } else {
                    self.error = error.localizedDescription + " Reconcile submission to check whether it was accepted."
                }
                isRunning = false
                _ = await save()
            }
        }
    }

    private func observe(_ id: String) async {
        defer { isRunning = false }
        var failures = 0
        while !Task.isCancelled {
            do {
                let run = try await transport.fetch(id)
                update(run)
                error = nil
                failures = 0
                _ = await save()
                if run.status.isTerminal { return }
            } catch {
                self.error = "Reconnecting to the original machine: \(error.localizedDescription)"
                failures += 1
                if failures >= 10 { return }
            }
            try? await Task.sleep(for: .seconds(failures == 0 ? 1 : 3))
        }
    }

    func reconnect() {
        guard !isRunning, let latest, !latest.status.isTerminal else { return }
        isRunning = true
        operation = Task { await observe(latest.id) }
    }

    func stop() {
        stopRequested = true
        guard pending == nil, let latest, !latest.status.isTerminal else { return }
        Task {
            do { try await cancelAccepted(latest.id) }
            catch { self.error = error.localizedDescription }
        }
    }

    private func cancelAccepted(_ id: String) async throws {
        update(try await transport.stop(id))
        _ = await save()
    }

    func promote() {
        guard let latest, latest.status == .completed, !isRunning, !isPromoting else { return }
        isPromoting = true
        Task {
            defer { isPromoting = false }
            do {
                let result = try await transport.promote(latest.id)
                update(result)
                _ = await save()
                if let pane = result.promotedPaneID { transport.openAgent(pane) }
            } catch { self.error = error.localizedDescription }
        }
    }

    func openAgent() {
        if let pane = latest?.promotedPaneID { transport.openAgent(pane) }
    }

    func newQuestion() {
        guard !isRunning, !isPromoting, pending == nil, latest?.status.isTerminal != false else { return }
        turns = []
        error = nil
        draft = ""
        Task { _ = await save() }
    }

    func refreshContext() {
        guard canSend else { return }
        context = currentContext
        context.snapshotId = UUID().uuidString
        context.capturedAt = Date.now.ISO8601Format()
        Task { _ = await save() }
    }

    func addContext(_ text: String) {
        guard canSend else { return }
        guard text.utf8.count <= 16 * 1024, context.items.count < 16 else {
            error = "Attach at most 16 items, each smaller than 16 KiB."
            return
        }
        context.items.append(.init(id: UUID().uuidString, kind: "text.v1", label: "Added context", text: text))
        context.snapshotId = UUID().uuidString
        context.capturedAt = Date.now.ISO8601Format()
        Task { _ = await save() }
    }

    func removeContext(_ id: String) {
        guard canSend else { return }
        context.items.removeAll { $0.id == id }
        context.snapshotId = UUID().uuidString
        context.capturedAt = Date.now.ISO8601Format()
        Task { _ = await save() }
    }

    func saveDraft() { Task { _ = await save() } }

    @discardableResult private func save() async -> Bool {
        do {
            try await persistence.save(.init(context: context, draft: draft, turns: turns, pending: pending, selectedModel: selectedModel))
            return true
        } catch {
            self.error = "Could not save this question: \(error.localizedDescription)"
            return false
        }
    }

    private func update(_ run: HeadlessAgentRun) {
        if let index = turns.firstIndex(where: { $0.id == run.id }) { turns[index] = run }
        else { turns.append(run) }
    }
}
